src/server/db_slice.cc

```diff
@@ -364,86 +366,122 @@ DbSlice::AddOrFindResult& DbSlice::AddOrFindResult::operator=(ItAndUpdater&& o)
+  auto res = FindInternal(cntx, key, req_obj_type, UpdateStatsMode::kUpdateMutableStats, load_mode);
+  if (!res.ok()) {
+    return res.status();
   }
 
-  PreUpdate(cntx.db_index, it);
-  return {{it, exp_it,
+  PreUpdate(cntx.db_index, res->it);
+  return {{res->it, res->exp_it,
            AutoUpdater({.action = AutoUpdater::DestructorAction::kRun,
                         .db_slice = this,
                         .db_ind = cntx.db_index,
-                        .it = it,
+                        .it = res->it,
                         .key = key})}};
 }
 
 DbSlice::ItAndExpConst DbSlice::FindReadOnly(const Context& cntx, std::string_view key) {
-  auto res = FindInternal(cntx, key, FindInternalMode::kUpdateReadStats);
-  return {res.it, res.exp_it};
+  auto res = FindInternal(cntx, key, std::nullopt, UpdateStatsMode::kUpdateReadStats,
+                          LoadExternalMode::kDontLoadExternal);
+  return {res->it, res->exp_it};
 }
 
 OpResult<PrimeConstIterator> DbSlice::FindReadOnly(const Context& cntx, string_view key,
                                                    unsigned req_obj_type) {
-  auto it = FindReadOnly(cntx, key).it;
-
-  if (!IsValid(it))
-    return OpStatus::KEY_NOTFOUND;
-
-  if (it->second.ObjType() != req_obj_type) {
-    return OpStatus::WRONG_TYPE;
+  auto res = FindInternal(cntx, key, req_obj_type, UpdateStatsMode::kUpdateReadStats,
+                          LoadExternalMode::kDontLoadExternal);
+  if (res.ok()) {
+    return {res->it};
   }
-
-  return {it};
+  return res.status();
 }
 
-DbSlice::ItAndExp DbSlice::FindInternal(const Context& cntx, std::string_view key,
-                                        FindInternalMode mode) {
-  DbSlice::ItAndExp res;
+OpResult<PrimeConstIterator> DbSlice::FindAndFetchReadOnly(const Context& cntx,
+                                                           std::string_view key,
+                                                           unsigned req_obj_type) {
+  auto res = FindInternal(cntx, key, req_obj_type, UpdateStatsMode::kUpdateReadStats,
+                          LoadExternalMode::kLoadExternal);
+  if (res.ok()) {
+    return {res->it};
+  }
+  return res.status();
+}
 
-  if (!IsDbValid(cntx.db_index))
-    return res;
+OpResult<DbSlice::ItAndExp> DbSlice::FindInternal(const Context& cntx, std::string_view key,
+                                                  std::optional<unsigned> req_obj_type,
+                                                  UpdateStatsMode stats_mode,
+                                                  LoadExternalMode load_mode) {
+  DCHECK(IsDbValid(cntx.db_index));
 
+  DbSlice::ItAndExp res;
   auto& db = *db_arr_[cntx.db_index];
   res.it = db.prime.Find(key);
-  FiberAtomicGuard fg;
-  if (!IsValid(res.it)) {
-    switch (mode) {
-      case FindInternalMode::kUpdateMutableStats:
+
+  auto update_stats_on_miss = [&](UpdateStatsMode stats_mode) {
```

Comment:
Instead of defining a lambda, I'd use an `absl::Cleanup` to update the misses/mutations. Then, at the end when we know we don't return an error, I'd `Cancel()` this cleanup and update the proper stats (like you already do)

Reply:
sounds good